[SelectionDAG] Fix bug related to demanded bits/elts for BITCAST

bjope opened the pull request:
When we have a BITCAST and the source type is a vector with smaller elements compared to the destination type, then we need to demand all the source elements that make up the demanded elts for the result when doing recursive calls to SimplifyDemandedBits, SimplifyDemandedVectorElts and SimplifyMultipleUseDemandedBits. Problem is that those simplifications are allowed to turn non-demanded elements of a vector into POISON, so unless we demand all source elements that make up the result there is a risk that the result would be more poisonous (even for demanded elts) after the simplification.

The patch fixes some bugs in SimplifyMultipleUseDemandedBits and SimplifyDemandedBits for situations when we did not consider the problem described above. Now we make sure that we also demand vector elements that "must not be turned into poison" even if those elements correspond to bits that does not need to be defined according to the DemandedBits mask.

Fixes #138513

bjope committed:
[SelectionDAG] Fix bug related to demanded bits/elts for BITCAST

When we have a BITCAST and the source type is a vector with smaller
elements compared to the destination type, then we need to demand
all the source elements that make up the demanded elts for the
result when doing recursive calls to SimplifyDemandedBits,
SimplifyDemandedVectorElts and SimplifyMultipleUseDemandedBits.
Problem is that those simplifications are allowed to turn non-demanded
elements of a vector into POISON, so unless we demand all source
elements that make up the result there is a risk that the result
would be more poisonous (even for demanded elts) after the
simplification.

The patch fixes some bugs in SimplifyMultipleUseDemandedBits and
SimplifyDemandedBits for situations when we did not consider the
problem described above. Now we make sure that we also demand vector
elements that "must not be turned into poison" even if those elements
correspond to bits that does not need to be defined according to
the DemandedBits mask.

Fixes #138513

bjope commented:
Doing a simple fix like this result in lots of lit test diffs. Unfortunately, I guess mostly regressions.

It could be possible to pass around a "MustNotMakeUndemandedEltsMorePoisonous" flag (or bitmask) in the recursive calls to SimplifyDemanded*, and set that to true for the BITCAST cases. It would allow us to do most simplifications even for BITCAST when we know that the bits from some source elements aren't used (although they must not be turned into poison). Such a hack would reduce the amount of lit test diffs and regressions. But we need to make a lot of boiler plate code changes, as well as finding all places where undemanded elements can be simplified as poison.

Another option might be to introduce new DAG combines etc to reduce amount of regressions that way. No idea how complicated that would be.

RKSimon reviewed (commented):
We have isGuaranteedNotToBeUndefOrPoison - can that be used somehow? 

`<edit>` I assume not as we're trying to not ADD poison as opposed to handle existing poison

bjope committed:
Adding DoNotPoisonEltMask to SimplifyDemandedVectorElts

bjope commented:
> We have isGuaranteedNotToBeUndefOrPoison - can that be used somehow?
> 
> `<edit>` I assume not as we're trying to not ADD poison as opposed to handle existing poison

One thing I experimented with was to make sure the SimplifyDemanded* functions avoided to make undemanded elements more poisonous. And then I would need to guard lots of things by checking isGuaranteedNotToBeUndefOrPoison. I could for example eliminate an INSERT_VECTOR_ELT if the inserted element wasn't demanded, but only if the same element was guaranteed not to be poison in the source vector. I think I got pretty far with such an approach, but unclear how many places I had missed. I must say that the recursions in SimplifyDemandedBits is a bit complicated to follow. It may recurse into both SimplifyDemandedBits, SimplifyMultipleUseDemandedBits, SimplifyMultipleUseDemandedVectorElts, SimplifyDemandedVectorElts, SimplifyDemandedBitsForTargetNode and SimplifyDemandedVectorEltsForTargetNode (and probably some more). So it is a quite big nest of code that is impacted if we want to change something like that.

RKSimon commented:
So this is what we're fighting I guess? https://alive2.llvm.org/ce/z/rqp6yU

```
----------------------------------------
define i8 @src1(<2 x i4> %a0, i4 %a1) {
#0:
  %v = insertelement <2 x i4> %a0, i4 %a1, i32 0
  %b = bitcast <2 x i4> %v to i8
  %a = and i8 %b, 240
  ret i8 %a
}
=>
define i8 @tgt1(<2 x i4> %a0, i4 %a1) {
#0:
  %b = bitcast <2 x i4> %a0 to i8
  %a = and i8 %b, 240
  ret i8 %a
}
Transformation doesn't verify!
----------------------------------------
define i8 @src2(<2 x i4> %a0, i4 %a1) {
#0:
  %v = insertelement <2 x i4> %a0, i4 %a1, i32 0
  %b = bitcast <2 x i4> %v to i8
  %a = and i8 %b, 240
  ret i8 %a
}
=>
define i8 @tgt2(<2 x i4> %a0, i4 %a1) {
#0:
  %v = insertelement <2 x i4> %a0, i4 0, i32 0
  %b = bitcast <2 x i4> %v to i8
  ret i8 %b
}
```

bjope commented:
> So this is what we're fighting I guess? https://alive2.llvm.org/ce/z/rqp6yU
> 

Yes, I think that is same thing (but wit IR and not selection DAG).

Often I guess it doesn't cause problems. Downstream we started to see miscompiles after the patch (https://github.com/llvm/llvm-project/pull/136701) that introduced the "zext of poison into poison" fold. The custom lowering of BITCAST for our downstream target would use a zext/shift/or kind of pattern to lower the bitcasted vector, and while we've been lucky that it worked in the past (even with a poisoned element before the bitcast) we now just got poison back.

I'm still not fully clear on how to proceed here. This patch, as is, does leave a lot to think about. The experiment with passing around "not really demanded elements but we must not make undemanded elements more poisonous" got a bit messy after awhile. I could perhaps add it to this patch anyway to show what I've been trying to do.

bjope committed:
Add DoNotPoisonEltMask to SimplifyDemandedVectorEltsForTargetNode

I don't know much about those X86 instructions.

Tried to handle DoNotPoisonEltMask for several X86ISD nodes,
but there is a bunch of situation when I just treated the
DoNotPoisonEltMask as being included in DemandedElts.

If taking X86ISD::VBROADCAST as an example it is a bit tricky.
That one can use widenSubVector to create a vector with UNDEF
elements. But that helper is using getNode(ISD::INSERT_SUBVECTOR)
which actually may return a vector with POISON elements. So then
we need to consider also the "do not poison elements" as being
demanded. If we fix https://github.com/llvm/llvm-project/issues/141034
then maybe we can trust widenSubVector to not return a more
poisonous vector, and simplify based on just DemandedElts.

bjope commented:
Added some commits showing the attempt of adding a "Do Not Poison Elts" kind of mask to SimplifyDemandedVectorElts (as well as the target node version in a separate step). That feels a bit tricky but reduced the amount of diffs a bit.

arsenm commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -4192,6 +4192,15 @@ class TargetLowering : public TargetLoweringBase {
   ///    with TLO.New will be incorrect when this parameter is true and TLO.Old
   ///    has multiple uses.
+  /// Vector elements that aren't demanded can be turned into poison unless the
+  /// corresponding bi in the \p DoNotPoisonEltMask is set.
```
```suggestion
  /// corresponding bit in the \p DoNotPoisonEltMask is set.
```

bjope commented:
Continued in an other PR as per previews comment.
